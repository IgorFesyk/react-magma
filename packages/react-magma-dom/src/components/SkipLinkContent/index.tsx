import * as React from 'react';

import { TARGET_ID } from '../SkipLink';

export interface SkipLinkContentProps
  extends React.HTMLAttributes<HTMLDivElement> {
  testId?: string;
}

export const SkipLinkContent: React.FunctionComponent<SkipLinkContentProps> = ({
  children,
  testId,
}: SkipLinkContentProps) => (
  <div data-testid={testId} id={TARGET_ID} tabIndex={-1} style={{ outline: 0 }}>
    {children}
  </div>
);
