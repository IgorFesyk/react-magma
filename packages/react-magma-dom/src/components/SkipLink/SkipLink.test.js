import React from 'react';
import { axe } from 'jest-axe';
import { SkipLink } from '.';
import { SkipLinkContent } from '../SkipLinkContent';

import { render, fireEvent } from 'react-testing-library';

describe('SkipLink', () => {
  it('should render the skip link component', () => {
    const { container } = render(<SkipLink />);
    const button = container.querySelector('button');

    expect(button).toBeInTheDocument();
    expect(button.innerHTML).toEqual('Skip Navigation');
    expect(button).toHaveStyleRule('background', '#006298');
    expect(button).toHaveStyleRule('color', '#FFFFFF');
    expect(button).toMatchSnapshot();
  });

  it('should render the skip link component with custom text', () => {
    const TEXT = 'Test text';
    const { container } = render(<SkipLink buttonText={TEXT} />);
    const button = container.querySelector('button');

    expect(button).toBeInTheDocument();
    expect(button.innerHTML).toEqual(TEXT);
  });

  it('should put focus on the h1 in the main content area when the button is clicked', () => {
    const { container } = render(
      <>
        <SkipLink />
        <SkipLinkContent>
          <h1>Test</h1>
        </SkipLinkContent>
      </>
    );
    const button = container.querySelector('button');
    const heading = container.querySelector('h1');

    expect(heading).not.toHaveFocus();
    fireEvent.click(button);
    expect(heading).toHaveFocus();
  });

  it('should put focus on the main content area when the button is clicked if they content area does not have an h1', () => {
    const { container } = render(
      <>
        <SkipLink />
        <SkipLinkContent />
      </>
    );
    const button = container.querySelector('button');
    const contentDiv = container.querySelector('#reactMagmaMainContent');

    expect(contentDiv).not.toHaveFocus();
    fireEvent.click(button);
    expect(contentDiv).toHaveFocus();
  });

  it('should not move focus if there is no content area', () => {
    const { container } = render(<SkipLink />);
    const button = container.querySelector('button');

    button.focus();
    fireEvent.click(button);
    expect(button).toHaveFocus();
  });

  it('should render the skip link button with specified color and variant', () => {
    const { container } = render(
      <SkipLink color="success" variant="outline" />
    );
    const button = container.querySelector('button');

    expect(button).toHaveStyleRule('background', 'rgba(0,0,0,0)');
    expect(button).toHaveStyleRule('color', '#3A8200');
  });

  it('should render the skip link button the correct colors for an inverse button', () => {
    const { container } = render(<SkipLink inverse />);
    const button = container.querySelector('button');

    expect(button).toHaveStyleRule('background', '#FFFFFF');
    expect(button).toHaveStyleRule('color', '#006298');
  });

  it('should render the skip link button specified position top and left attributes', () => {
    const { container } = render(
      <SkipLink positionLeft={86} positionTop={99} />
    );
    const button = container.querySelector('button');

    expect(button).toHaveStyleRule('left', '86px', {
      target: ':focus'
    });
    expect(button).toHaveStyleRule('top', '99px', {
      target: ':focus'
    });
  });

  it('Does not violate accessibility standards', () => {
    const { container } = render(<SkipLink />);

    return axe(container.innerHTML).then(result => {
      return expect(result).toHaveNoViolations();
    });
  });
});
