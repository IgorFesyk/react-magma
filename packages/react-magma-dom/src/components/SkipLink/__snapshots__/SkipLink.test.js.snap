// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipLink should render the skip link component 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 5px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-family: "Open Sans",sans-serif;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  line-height: 1;
  margin: 5px;
  min-width: 5.625em;
  overflow: hidden;
  position: relative;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-transition: background 0.35s,color 0.35s;
  transition: background 0.35s,color 0.35s;
  vertical-align: middle;
  touch-action: manipulation;
  white-space: nowrap;
  font-size: 14px;
  font-weight: 600;
  height: 37px;
  padding: 0 15px;
  width: auto;
  background: #006298;
  border: 0;
  border-color: #006298;
  color: #FFFFFF;
  left: -9999px;
  position: fixed;
  top: -9999px;
}

.emotion-0:not(:disabled):focus {
  outline: 2px dotted #027EE1;
  outline-offset: 3px;
}

.emotion-0:not(:disabled):hover,
.emotion-0:not(:disabled):focus {
  background: #004165;
  color: #FFFFFF;
}

.emotion-0:not(:disabled):after {
  border-radius: 50%;
  content: '';
  height: 32px;
  left: 50%;
  opacity: 0;
  position: absolute;
  padding: 50%;
  top: 18px;
  -webkit-transform: translate(-50%,-50%) scale(1);
  -ms-transform: translate(-50%,-50%) scale(1);
  transform: translate(-50%,-50%) scale(1);
  -webkit-transition: opacity 1s,-webkit-transform 0.5s;
  -webkit-transition: opacity 1s,transform 0.5s;
  transition: opacity 1s,transform 0.5s;
  width: 32px;
  background: #FFFFFF;
}

.emotion-0:not(:disabled):active {
  background: #002032;
  color: #FFFFFF;
}

.emotion-0:not(:disabled):active:after {
  opacity: 0.4;
  -webkit-transform: translate(-50%,-50%) scale(0);
  -ms-transform: translate(-50%,-50%) scale(0);
  transform: translate(-50%,-50%) scale(0);
  -webkit-transition: -webkit-transform 0s;
  -webkit-transition: transform 0s;
  transition: transform 0s;
}

.emotion-0:focus {
  left: 10px;
  top: 10px;
  z-index: 99999;
}

<a
  class="emotion-0"
  color="primary"
  href="#reactMagmaMainContent"
  variant="solid"
>
  Skip Navigation
</a>
`;
